### DISCUSSION TITLE
lab3中的中断跳转与清空

### DISCUSSION THREAD

**Discussion by linyn-zero** [SYSU-SCC/yatcpu-docs#15] [General]
我发现lab3的标准测试对于中断发生时，应该怎么清空流水线寄存器，管得很松，不过都很有意思。

Final实验。情景：当CLINT要执行中断跳转（mepc会保存当前PC的值），且此时CPU上没有跳转/分支指令，没有数据冒险，没有Load指令。
也就是说，IF模块的指令被替换成了中断处理程序的第一条指令，而与此同时ID/EX/MEM/WB模块的指令则需要照常执行，对不对？
也就是说，中断跳转根本不需要清空任何流水线寄存器对不对？
而由CPU中跳转/分支指令产生的跳转，需要清空一些流水线寄存器，因跳转清空时需要判断是哪种跳转对不对？
但是判不判定都能过欸……

然后是forwarding实验。贴上我实验报告的一部分：
（定义“依赖”为`rs_id === rd_ex`）
现分析在同时满足阻塞和清空的条件的情况下，Control单元应该采取的行为：

1. 清空的条件是EX模块的`if_jump_flag`信号为真，即（CLINT跳转信号 或 EX模块为跳转指令 或  EX模块为分支指令且分支判断）为真。
2. 阻塞的条件是ID模块的rs等于EX模块的rd，且EX模块为Load指令。

因此两者同时发生时只有一种可能——即此时EX模块为Load指令，ID模块的指令依赖Load指令，CLINT发出跳转信号。（“依赖”定义为`rs_id === rd_ex`）
那么这种情况Control单元究竟该阻塞，还是清空，还是同时阻塞清空呢？（虽然测试都能过，不过我在CLINT的代码中发现，助教大神已经考虑到了这个问题。）见源码图：
![image](https://user-images.githubusercontent.com/112923198/212943304-bd825faa-b00d-4858-93a4-7858dd5a91dc.png)
这种情况下，中断处理结束后将返回到此时的依赖指令，也就是说，此时Control单元应该将依赖指令清空且不阻塞。
不过我还是不明白为什么CLINT此时要保存ID的指令，明明ID的指令依赖于EX指令，在两者中间执行中断处理程序可能会导致ID指令的RegData错误吧？望指教。

**Comment by huangcb01** [comment]
同学你好，Lab3 是我设计的。其实在流水线中加入中断是一件挺麻烦的事，我在写代码时也在中断这块遇到了比较多的问题，加上中断是 Lab2 的主题，所以我在 Lab3 的实验设计中就刻意避开了中断这个话题，但是在心里又怕同学们做这种填空式的题目时只见树木，不见森林。所以我非常高兴你能够对基础作业要求以外的部分产生兴趣并进行探索。

有兴趣的话，你还可以看一下 [YatCPU 原始的代码](https://github.com/howardlau1999/yatcpu) 和 [文档](https://yatcpu.sysu.tech/tutorial/arch/)。最初的三级流水线 YatCPU 是由 @howardlau1999 学长实现的，我在他的基础上又实现了五级流水线。再后来，我与另外两位同学分别修改 YatCPU 的代码，设计了这几个实验。（在主仓库的五级流水线 YatCPU 中，因为 IF 段通过总线取指需要多个周期，掩盖了阻塞和旁路中的几个 bug，后来我在写 Lab3 的代码时发现并改过来了，但是主仓库中的代码还没有修正，我应该会在寒假把这个坑填上……）

我一开始读三级流水线 YatCPU 的代码就觉得有些地方实现得比较繁琐，所以在我写代码的时候一直秉持着“放宽约束，简化逻辑的原则”，希望可以简化底层逻辑，节省硬件资源，减小组合逻辑延迟，同时也让代码更简洁。当然，这样也可能更容易出现 bug，在未来拓展其他指令的时候也可能会带来麻烦。但我还是把这个原则带到了 Lab3 的设计中：原来的 CLINT 是一个状态机，在处理中断时需要多个周期，对 CSR 逐个写入，我和其他设计实验的同学进一步把它简化为在 1 个周期内同时写入所有需要写入的 CSR。这样，中断对流水线的影响可以用一句话概括：中断发生时，相当于在处理跳转/分支的流水段上叠加了一条跳转指令。不知道这是否有解决你的第一个疑惑？下面我会做详细解释。

-----------------------------------------

下面进入正题。

1. 你理解的“IF模块的指令被替换成了中断处理程序的第一条指令”不准确。

	这是 FiveStageFinal 的译码段最后几行代码（不知道你们作业截止了没有，我是不是透露了部分答案哈哈）：

	```scala
	  val instruction_jump_address = io.ex_immediate + Mux(opcode === Instructions.jalr, reg1_data, io.instruction_address)
	  io.clint_jump_flag := instruction_jump_flag
	  io.clint_jump_address := instruction_jump_address
	  io.if_jump_flag := io.interrupt_assert || instruction_jump_flag
	  io.if_jump_address := Mux(io.interrupt_assert,
	    io.interrupt_handler_address,
	    instruction_jump_address
	  )
	```

	可以看到，CLINT 发来的 `interrupt_assert` 经过 ID 的处理，已经与普通的跳转指令没什么差别了，都是向 IF 和 Control 发送 `if_jump_flag` 和 `if_jump_address`，然后 IF 和 Control 就会当作这里有一条跳转指令来处理，Control 会向处理跳转指令那样，清空 IF 段，然后 IF 会从 `if_jump_address` 处开始取指。但是，原本 ID 段上的指令并没有被破坏，所以说是在 ID 段原本的指令上“叠加了一条跳转指令”。

2. 这是由软件负责解决的。

	其实，不仅仅是在相邻两条存在依赖的指令中间发生中断，即使是在隔着若干条指令的两条存在依赖的指令中间发生中断，中断服务子程序完全有可能“篡改”寄存器组中的数据，导致原本的程序出错。这完全取决于你的中断服务子程序是怎么写的。因此，一般在中断服务子程序开始时需要**保存现场**，其中就包括把需要用到的寄存器入栈（暂时保存到内存里），最后再**恢复现场**，把寄存器的值出栈。当然，如果你是用高级语言写的程序，编译器已经帮你把事情做好了。下图是反汇编 `tetris` 得到的汇编代码：

	![image](https://user-images.githubusercontent.com/78298502/213062682-5366d812-00a6-4b28-9296-73d68bde4ebf.png)

	另外，`jumpping` 表示在上一周期在 EX 段是否有一次跳转（可以是跳转/分支指令，也可以是中断跳转），如果有，那么现在 ID 里的指令地址其实是被清空了的，所以应该保存 IF 中的指令地址。所以 `io.jump_flag` 并不是没有意义的。

---------------------------------------------

其实我已经有几个月没有接触 YatCPU 了，所以我的理解可能也不一定准确。如果还有任何问题，欢迎继续讨论~

**Comment by linyn-zero** [comment]
谢谢学长，我知道我final的疑惑出于哪里了——我对清空流水线寄存器的效果没有理解到位。在final里，指令跳转与中断跳转的效果是一样的，即在ID段发出一条跳转指令。我们在此时应该清空if2id，是清空原来的IF段的结果，是不会影响当前ID段的指令的。（效果则是IF段下一周期是跳转目标指令，ID段下一周期是气泡，EX段下一周期是原来的ID段指令）

forwarding的问题在于我忘记了中断处理会保护现场，罪过罪过……不过jumping在实际效果上，它服务于的instruction_address也使用了jump_flag来Mux，使得jumping中的jump_flag没有效果，但是我理解jumping中的jump_flag的逻辑意义

**Reply by huangcb01** [reply]
请注意 `jumpping` 是一个寄存器，它保存的是上一周期的 `io.jump_flag || io.ex_interrupt_assert`，而 `instruction_address` 中 Mux 的是当前周期的 `io.jump_flag`。

**Reply by linyn-zero** [reply]
我悟了，谢谢学长！

**Comment by linyn-zero** [comment]
我在写实验报告Final的Control时，又遇到了一个问题
![image](https://user-images.githubusercontent.com/112923198/213095131-cdff9dc8-6544-4778-938b-508391e3409f.png)
首先我不确定自己在3、4条的判断是否正确，如果正确，第4条似乎又实现不了

**Reply by huangcb01** [reply]
如果我没有记错，Final 中需要阻塞的非跳转/分支指令的数据冲突应该就是 ID 段的指令依赖于 EX 段的 Load 指令吧，这样，让 ID 段的指令阻塞一个周期，等原本 ID 段的指令到 EX 段时，Load 指令已经到了 WB 段，可以通过旁路获取所需数据。

需要阻塞的跳转/分支指令的数据冲突应该是指 ID 段的分支/跳转指令依赖于 EX 段的指令或 MEM 段的 Load 指令，都是把 ID 段的分支/跳转指令阻塞一个周期。（特别地，如果 ID 段的分支/跳转依赖于 EX 段的 Load 指令，满足了第一个检测条件，阻塞一个周期，然后又满足第二个检测条件，再阻塞一个周期。）

---------------------------------------------------

上面的分析中我没有考虑到中断，下面进入正题。

**第三条**：ID 段的指令依赖于 EX 段的 Load 指令，并且此时发生中断，那么不会跳转（不会清空），而是阻塞 ID 和 CLINT，没错，CLINT 也被阻塞了，这是因为：（如下图）外界来的中断信号 `IntFlag` 是经过 IF2ID 暂存的，而 IF2ID 被阻塞了；而如果是内部中断，内部中断是由 ID 段的指令引发的，ID 段的指令也被阻塞了。所以，在阻塞 IF 和 ID 的同时其实也阻塞了 CLINT。所以，中断的处理还是在中断发生时 ID 段的指令上叠加了一个跳转，只不过因为 ID 段指令上的数据冲突，两者都在 ID 段阻塞了一个周期而已。

![图片](https://yatcpu.sysu.tech/labs/images/five_stage_pipelined_CPU_structure.png)

**第四条**：应该也不难理解了，中断跟着跳转/分支指令一起阻塞一或二个周期，然后跳转/分支指令的数据冲突解决了，可以算出目标地址了，把目标地址存进 `mepc`，实际跳转地址设为中断处理子程序入口，然后再清空 IF2ID。

-----------------------

不知道我的解释有没有问题，有些情况我在写代码的时候可能都没有去仔细想过，我现在也是想了很久才捋清楚。如果还有问题，欢迎继续提出。

**Reply by linyn-zero** [reply]
太妙了学长，没想到把IntFlag先送到if2id就能解决这么多问题！您究竟是怎么设计出如此棒的CPU的？

**Reply by huangcb01** [reply]
哈哈，其实这不是我设计的，原本的三级流水线 CPU 里就有把 `IntFlag` 先送到 IF2ID，我当时也是照抄过来，不明白为什么要这样做。其实还有其他很多复杂的情况在设计的时候根本没有想到，都是 CPU 跑程序的时候遇到 bug 再改过来的。
